eventsourcing/src/main/java/org/axonframework/eventsourcing/AnnotationBasedEventStateApplier.java

```diff
@@ -0,0 +1,136 @@
+     *                            {@link org.axonframework.eventhandling.EventMessage EventMessages}
+     */
+    public AnnotationBasedEventStateApplier(@Nonnull Class<M> modelType,
+                                            @Nonnull MessageTypeResolver messageTypeResolver) {
+        this(modelType, ClasspathParameterResolverFactory.forClass(modelType), messageTypeResolver);
+    }
+
+    /**
+     * Initialize a new {@link AnnotationBasedEventStateApplier}.
+     *
+     * @param modelType           The type of model this instance will handle state changes for.
+     * @param parameterResolverFactory the strategy for resolving handler method parameter values
+     * @param messageTypeResolver      The {@link MessageTypeResolver} resolving the
+     *                                 {@link org.axonframework.messaging.MessageType types} for
+     *                                 {@link org.axonframework.eventhandling.EventMessage EventMessages}
+     */
+    private AnnotationBasedEventStateApplier(
+            @Nonnull Class<M> modelType,
+            @Nonnull ParameterResolverFactory parameterResolverFactory,
+            @Nonnull MessageTypeResolver messageTypeResolver) {
+        this(parameterResolverFactory,
+             ClasspathHandlerDefinition.forClass(modelType),
+             messageTypeResolver);
+    }
+
+    /**
+     * Initialize a new {@link AnnotationBasedEventStateApplier}.
+     *
+     * @param parameterResolverFactory the strategy for resolving handler method parameter values
+     * @param handlerDefinition        the handler definition used to create concrete handlers
+     * @param messageTypeResolver      The {@link MessageTypeResolver} resolving the
+     *                                 {@link org.axonframework.messaging.MessageType types} for
+     *                                 {@link org.axonframework.eventhandling.EventMessage EventMessages}
+     */
+    public AnnotationBasedEventStateApplier(@Nonnull ParameterResolverFactory parameterResolverFactory,
+                                            @Nonnull HandlerDefinition handlerDefinition,
+                                            @Nonnull MessageTypeResolver messageTypeResolver) {
+        requireNonNull(parameterResolverFactory, "ParameterResolverFactory may not be null");
+        requireNonNull(handlerDefinition, "HandlerDefinition may not be null");
+        requireNonNull(messageTypeResolver, "MessageTypeResolver may not be null");
+        this.parameterResolverFactory = parameterResolverFactory;
+        this.handlerDefinition = handlerDefinition;
+        this.messageTypeResolver = messageTypeResolver;
+    }
+
+    @Override
+    public M apply(M model, @Nonnull EventMessage<?> event) {
+        requireNonNull(model, "Model may not be null");
+        requireNonNull(event, "Event Message may not be null");
+
+        // todo: should we create a new instance of AnnotationEventHandlerAdapter for each event?
```

Comment:
I think we have to in the current setup, as the `AnnotationEventHandlerAdapter` works on an instance instead of a `Class`. A future installment could conjure up the event handlers based on the `Class` for sure. But that will have to wait until the `AnnotationEventHandlerAdapter` is replaced with something like the `EventHandlingComponent` I think.